bundles/alpha.codegen/src/alpha/codegen/isl/MemoryUtils.xtend

```diff
@@ -0,0 +1,127 @@
+	 * less than that point.
+	 * We will compute this by building the set of such points, then using the Barvinok
+	 * library to compute the cardinality of such a set.
+	 * This will return a piecewise quasi-affine polynomial.
+	 * 
+	 * Note: the polynomial returned may look like it's undefined for the lex min of the given set.
+	 * However, this is not actually the case.
+	 * These polynomials are defined over the universe, and will evaluate to zero
+	 * if the point you're asking for lies outside the domains of each piece.
+	 * In other words: there is an implicit "default" case that always evaluates to zero
+	 * if the point being evaluated lies outside all the domains of all the pieces of the polynomial.
+	 * 
+	 * For example, the set "[N] -> {[i]: 0 <= i <= N}" (which forms a line from 0 to N)
+	 * would return the polynomial "[N,i] -> { i : 0 < i <= N }".
+	 * Per the previous paragraph, this is still valid for i=0,
+	 * as it falls under the implicit extra case that always evaluates to zero.
+	 * 
+	 * The intended use case of this is for determining the index of a point 
+	 * within a linearized array of our set. This produces a 0-based indexing.
+	 * Per the second paragraph, it may look like this polynomial is not defined for index 0,
+	 * but it actually is due to the implicit default case always evaluating to 0
+	 * if the point you're evaluating at is not in any of the domains of the pieces.
+	 * 
+	 * @param domain The domain to get the ranking polynomial for. This set is not destroyed ("isl_keep").
+	 * @returns A piecewise quasi-affine polynomial that can be used to compute the rank of a point within the given domain.
+	 */
+	def static ISLPWQPolynomial rank(ISLSet domain) {
+		// At a high level, we will simply construct a set which represents
+		// all points lexicographically smaller or equal to some desired point.
+		// We intersect this with the original domain to get only the points we care about.
+		// The cardinality of this set will be a polynomial which indicates the rank of our desired point.
+		
+		// To get the rank of our desired point, we need to add that point as a parameter.
+		// We'll do this by simply moving the indices of our given domain to be parameters.
+		// This also moves the constraints to be parameter constraints.
+		// We then introduce new indices that represent the original ones.
+		// We don't need to re-add the original constraints, as intersecting
+		// this set with the original domain does this for us.
+
+		// The introduced parameters will all be named like "i0", "i1", etc.
+		// However, we don't want duplicate names, so we skip any that already exist.
+		// Note: lazy evaluation means this won't actually generate all Integer.MAX_VALUE names.
+		val existingNames = #[domain.paramNames, domain.indexNames].flatten.toArrayList
+		val names = (0 ..< Integer.MAX_VALUE)
+			.map["i" + it]
+			.reject[existingNames.contains(it)]
```

Comment:
Cool, this handles the case where the input set contains index names like “i0”, “i1”, etc.

Given this domain: `[N] -> { [i0] : N >= 11 and 0 <= i0 < N }`
* `extendedSet` —> `[N, i0] -> { [i1] : N >= 11 and 0 <= i0 < N }`
* `lessThan` —> `[N, i0] -> { [i1] : N >= 11 and 0 <= i0 < N and i1 < i0 }`
* returned cardinality —> `[N, i0] -> { i0 : N >= 11 and 0 < i0 < N }`

It basically just uses names that are not in the existingNames. And assuming this is all lazily evaluated, I’m assuming that it does not actually instantiate a list of size Integer.MAX_VALUE. Very nice!

Reply:
Yes, I've verified that the names are lazily evaluated, so we're not generating a list of about $2^{31}$ elements